File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def check(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        if (not root1) and (not root2):
            return True
        if (not root1) or (not root2):
            return False
        if root1.val==root2.val:
            return self.check(root1.left, root2.right) and self.check(root1.right, root2.left)
        return False

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.check(root.left, root.right)

        
        

File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        queue<pair<TreeNode*, TreeNode*>> q;
        q.push({root->left, root->right});
        while(!q.empty()) {
            pair<TreeNode*,TreeNode*> p = q.front();
            q.pop();
            if(!p.first && !p.second) continue;

            if(p.first && p.second && p.first->val == p.second->val) {
                q.push({p.first->right, p.second->left});
                q.push({p.first->left, p.second->right});
            } else {
                return false;
            }
        }
        return true;
        
    }
};